File: src/leetCode/backTrack/Problem039.java
package leetCode.backTrack;

import java.util.ArrayList;
import java.util.List;

/**
 * Given a set of candidate numbers (candidates) (without duplicates) and a target number (target),
 * find all unique combinations in candidates where the candidate numbers sums to target.
 * <p>
 * The same repeated number may be chosen from candidates unlimited number of times.
 * <p>
 * Note:
 * <p>
 * All numbers (including target) will be positive integers.
 * The solution set must not contain duplicate combinations.
 * create by stephen on 2018/6/28
 */
public class Problem039 {

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> result = new ArrayList<>();
        backTrack(result, candidates, target, 0, new ArrayList<>());
        return result;
    }

    private void backTrack(List<List<Integer>> result, int[] candidates, int target, int start, ArrayList<Integer> tempList) {
        if (target < 0) return;
        if (target == 0) result.add(new ArrayList<>(tempList));

        for (int i=start; i<candidates.length; ++i) {
            tempList.add(candidates[i]);
            backTrack(result, candidates, target-candidates[i], i, tempList);
            tempList.remove(tempList.size() - 1);
        }
    }
}

File: leetCode/backTrack/Problem039.cpp
#include <vector>

/**
 * Given an array of distinct integers candidates and a target integer target,
 * return a list of all unique combinations of candidates where the chosen
 * numbers sum to target. You may return the combinations in any order.
 *
 * The same number may be chosen from candidates an unlimited number of times.
 * Two combinations are unique if the frequency of at least one of the chosen
 * numbers is different.
 *
 * It is guaranteed that the number of unique combinations that sum up to target
 * is less than 150 combinations for the given input.
 */
void backtrack(std::vector<std::vector<int>>&, std::vector<int>&, int,
               std::vector<int>&, int);
std::vector<std::vector<int>> combinationSum(std::vector<int>& candidates,
                                             int target) {
  std::vector<std::vector<int>> result;
  std::vector<int> current;
  backtrack(result, candidates, 0, current, target);
  return result;
}

void backtrack(std::vector<std::vector<int>>& result,
               std::vector<int>& candidates, int idx, std::vector<int>& current,
               int remain) {
  if (remain < 0) return;
  if (remain == 0) {
    result.push_back(current);
    return;
  }
  for (int i = idx; i < candidates.size(); ++i) {
    current.push_back(candidates[i]);
    backtrack(result, candidates, i, current, remain - candidates[i]);
    current.pop_back();
  }
}